File: LeetCode/C/111-depth-first-search-二叉树的最小深度.cpp
/*
 * @lc app=leetcode.cn id=111 lang=cpp
 *
 * [111] 二叉树的最小深度
 */
#include "main.h"
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
  public:
    int minDepth(TreeNode *root) {
        if (!root)
            return 0;
        if (!root->left && !root->right)
            return 1;
        int res = INT_MAX;
        if (root->left)
            res = min(res, minDepth(root->left));
        if (root->right)
            res = min(res, minDepth(root->right));
        return res + 1;
    }
};
// @lc code=end

File: LeetCode/Go/111-depth-first-search-二叉树的最小深度.go
/*
 * @lc app=leetcode.cn id=111 lang=golang
 *
 * [111] 二叉树的最小深度
 */
package leetcode

import "math"

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var res = math.MaxInt64
	dfs(root, 1, &res)
	return res
}

func dfs(root *TreeNode, level int, res *int) {
	if root.Left == nil && root.Right == nil {
		*res = min(*res, level)
		return
	}
	if root.Left != nil {
		dfs(root.Left, level + 1, res)
	}
	if root.Right != nil {
		dfs(root.Right, level + 1, res)
	}
}

func min(i, j int) int {
	if i < j {
		return i
	} else {
		return j
	}
}
// @lc code=end

